dev_tools/modules.py

```diff
@@ -0,0 +1,161 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import dataclasses
+import os
+import sys
+from typing import List, Dict, Any
+
+_FOLDER = 'folder'
+_PACKAGE_PATH = 'package-path'
+
+
+@dataclasses.dataclass
+class Module:
+    root: str
+    raw_setup: Dict[str, Any]
+
+    name: str = dataclasses.field(init=False)
+    version: str = dataclasses.field(init=False)
+    top_level_packages: List[str] = dataclasses.field(init=False)
+
+    def __post_init__(self):
+        self.name = self.raw_setup['name']
+        if 'packages' in self.raw_setup:
+            self.top_level_packages = [p for p in self.raw_setup['packages'] if '.' not in p]
+        else:
+            self.top_level_packages = []
+        self.top_level_package_paths = [os.path.join(self.root, p) for p in self.top_level_packages]
+        self.version = self.raw_setup['version']
+
+
+def list_modules(
+    search_dir: str = os.path.join(os.path.dirname(__file__), '..'), include_parent: bool = False
+) -> List[Module]:
+    """Returns a list of python modules based defined by setup.py files.
+
+    Args:
+        include_parent: if true, a setup.py is expected in `search_dir`, and the corresponding
+        module will be included.
+        search_dir: the search directory for modules, by default the repo root.
+    Returns:
+          a list of `Module`s that were found, where each module `m` is initialized with `m.root`
+           relative to `search_dir`, `m.raw_setup` contains the dictionary equivalent to the
+           keyword args passed to the `setuptools.setup` method in setup.py
+    """
+
+    relative_folders = sorted(
+        [
+            f
+            for f in os.listdir(search_dir)
+            if os.path.isdir(os.path.join(search_dir, f))
+            and os.path.isfile(os.path.join(search_dir, f, "setup.py"))
+        ]
+    )
+    if include_parent:
+        parent_setup_py = os.path.join(search_dir, "setup.py")
+        assert os.path.isfile(
+            parent_setup_py
+        ), f"include_parent=True, but {parent_setup_py} does not exist."
+        relative_folders.append(".")
+
+    result = [
+        Module(root=folder, raw_setup=_parse_module(os.path.join(search_dir, folder)))
+        for folder in relative_folders
+    ]
+
+    return result
+
+
+def _parse_module(folder: str) -> Dict[str, Any]:
+    setup_args = {}
+    import setuptools
+
+    orig_setup = setuptools.setup
+    cwd = os.getcwd()
+
+    def setup(**kwargs):
+        nonlocal setup_args
+        setup_args = kwargs
```

Comment:
nit: might be simpler to update the outer dict, instead of assigning non-locally:
```suggestion
    def setup(**kwargs):
        setup_args.update(kwargs)
```